core/src/triggers.rs

```diff
@@ -8,82 +8,162 @@
+    ///
+    /// # Errors
+    /// - If trigger with such id already exists
+    pub fn add_pipeline_trigger(&self, trigger: Trigger<PipelineEventFilter>) -> Result<()> {
+        self.add_to(trigger, &self.pipeline_triggers)
+    }
+
+    /// Add trigger with [`TimeEventFilter`]
     ///
     /// # Errors
-    /// - If [`TriggerSet`] already contains a trigger with the same id.
-    /// It's the user's responsibility to first `Unregister` the `Trigger`.
-    pub fn add(&self, trigger: Trigger) -> Result<(), smartcontracts::Error> {
-        if self.0.contains_key(&trigger.id) {
+    /// - If trigger with such id already exists
+    pub fn add_time_trigger(&self, trigger: Trigger<TimeEventFilter>) -> Result<()> {
+        self.add_to(trigger, &self.time_triggers)
+    }
+
+    /// Add trigger with [`ExecuteTriggerEventFilter`]
+    ///
+    /// # Errors
+    /// - If trigger with such id already exists
+    pub fn add_by_call_trigger(&self, trigger: Trigger<ExecuteTriggerEventFilter>) -> Result<()> {
+        self.add_to(trigger, &self.by_call_triggers)
+    }
+
+    /// Add generic trigger to generic collection
+    ///
+    /// # Errors
+    /// - If trigger with such id already exists
+    fn add_to<F: Filter>(
+        &self,
+        trigger: Trigger<F>,
+        map: &DashMap<trigger::Id, Action<F>>,
+    ) -> Result<()> {
+        if self.contains(&trigger.id) {
             return Err(smartcontracts::Error::Repetition(
                 InstructionType::Register,
                 IdBox::TriggerId(trigger.id),
             ));
         }
-        self.0.insert(trigger.id, trigger.action);
 
+        map.insert(trigger.id.clone(), trigger.action);
+        self.ids.insert(trigger.id);
         Ok(())
     }
 
-    /// Get trigger by `id`
+    /// Get all contained triggers ids without particular order
+    pub fn ids(&self) -> Vec<trigger::Id> {
+        self.ids.iter().map(|entry| entry.clone()).collect()
+    }
+
+    /// Apply `f` to the trigger identified by `id`
     ///
     /// # Errors
     /// - If [`TriggerSet`] doesn't contain the trigger with the given `id`.
-    pub fn get(&self, id: &trigger::Id) -> Result<impl Deref<Target = Action> + '_, FindError> {
-        self.0.get(id).ok_or_else(|| FindError::Trigger(id.clone()))
+    pub fn inspect<F, R>(&self, id: &trigger::Id, f: F) -> std::result::Result<R, FindError>
+    where
+        F: Fn(&dyn ActionTrait) -> R,
+    {
+        if !self.contains(id) {
+            return Err(FindError::Trigger(id.clone()));
+        }
+
+        Ok(self
+            .data_triggers
+            .get(id)
+            .map(|entry| f(entry.value()))
+            .or_else(|| self.pipeline_triggers.get(id).map(|entry| f(entry.value())))
+            .or_else(|| self.time_triggers.get(id).map(|entry| f(entry.value())))
+            .or_else(|| self.by_call_triggers.get(id).map(|entry| f(entry.value())))
+            .expect("`TriggerSet` sub-sets doesn't have required id. This is a bug"))
```

Comment:
This `expect` is OK.

Reply:
I do think that we can avoid storing each key twice and having to search the entire `TriggerSet`. Consider:
1. generating `Id` from the `dashmaps` on-demand. In that case the invariant is enforced. 
2. storing an `enum` which tells you exactly which of the  `dashmaps` contained the value.